### PR TITLE
Add BatchNorm container + assorted housekeeping

### PR DESCRIPTION
**Pull request by jma127**
Integration tests show that adding batchnorm to the first linear layer improves accuracy ~2%, all else equal. This holds across affine and non-affine so is probably not just a result of the extra gamma/beta parameters.

One thing I'd like to know (I'll revise the PR as needed): how can I query the ``cudnn_enabled`` state? I assume using ``cuda_`` and hoping that ``cuda_`` implies ``cudnn_enabled`` is bad, but I was unable to find any cudnn related stuff here.

### CODE REVIEW

**Comment by ebetica** [comment]
Cudnn is enabled on the PyTorch side. Unfortunately, we can only check if it's enabled/acceptable in a more recent PR of PyTorch (https://github.com/pytorch/pytorch/pull/4749#event-1433559914). It should be automagically bound.

File: autograd.cpp [hunk 1]
```diff
@@ -1,4 +1,8 @@
+#include <algorithm>
 #include <cmath>
+#include <fstream>
```
Comment (ebetica) [inline autograd.cpp]: This include is already in autograd.h

File: autograd.cpp [hunk 2]
```diff
@@ -1,4 +1,8 @@
...
+#include <functional>
+#include <cereal/archives/binary.hpp>
```
Comment (ebetica) [inline autograd.cpp]: So is this one.

File: autograd.cpp [hunk 3]
```diff
@@ -207,6 +211,57 @@ variable_list Conv::forward(variable_list input) {
...
+    }
+    assert(channels > 1);
+  }
+
+  auto output = at::batch_norm(
+      input,
+      weight, bias,
+      running_mean, running_var,
+      train_, momentum_, eps_,
+      cuda_);  // TODO: figure out where to query cudnn usage
```
Comment (ebetica) [inline autograd.cpp]: `AT_CUDNN_ENABLED()`

File: autograd.h [hunk 4]
```diff
@@ -297,22 +297,50 @@ AUTOGRAD_CONTAINER_CLASS(Conv) {
...
+  Variable weight;
+  Variable bias;
+  Variable running_mean;
+  Variable running_var;
+
+ protected:
+  uint32_t num_features_;
+};
+
+class BatchNorm1d : public BatchNorm {
```
Comment (ebetica) [inline autograd.h]: Not much point in implementing BatchNorm1d and 2d actually, if they're treated the same.

File: autograd.cpp [hunk 5]
```diff
@@ -207,6 +211,57 @@ variable_list Conv::forward(variable_list input) {
...
+  }
+}
+
+variable_list BatchNorm::forward(variable_list inputs) {
+  auto& input = inputs[0];
+  auto& running_mean = (stateful_ ? this->running_mean : inputs[1]);
+  auto& running_var = (stateful_ ? this->running_var : inputs[2]);
+
+  if (train_) {
+    int64_t channels = input.sizes()[0];
```
Comment (ebetica) [inline autograd.cpp]: This block is just `input.numel() / input.size(1)`. Also, why is the assert necessary? In the case of numel() == 0 tensors? Why not just directly check for that then?

File: test.cpp [hunk 6]
```diff
@@ -609,67 +647,78 @@ std::map<std::string, void (*)()> constuct_tests() {
...
-       optim->zero_grad();
-       backward(loss);
-       optim->step(); 
-     }
-   }
+   auto forward = [&](Variable x) {
+     x = std::get<0>(at::max_pool2d(conv1->forward({x})[0], {2, 2})).clamp_min(0);
+     x = conv2->forward({x})[0];
+     x = drop2d->forward({x})[0];
+     x = std::get<0>(at::max_pool2d(x, {2, 2})).clamp_min(0);
```
Comment (ebetica) [inline test.cpp]: Add a batchnorm here (before the relu) to test for the 2d case?

**Comment by jma127** [comment]
The extra headers were from a merge conflict, sorry.

**Comment by ebetica** [comment]
Looks good to me! Thanks @jma127
